### PR TITLE
ENH: Input validation for sampling frequency in signal.filter (#14450)

### LINKED ISSUE #14450: Passing a numpy array as sampling frequency to signal.iirfilter causes unexpected error message
My issue is about the interface for filter design. When passing a numpy array instead of a single float as fs to iirfilter (and derived filters), the function crashes with the error message: "Must specify a single critical frequency Wn for lowpass or highpass filter". While clearly the function should crash, the error message seems misleading, as indeed only one critical frequency was passed to the function.

As far as I can tell, this issue arises from the following line, because the division of a single critical frequency by an array of sampling rates will return an array of critical frequencies.
https://github.com/scipy/scipy/blob/12b77ad2e2d8ee766566e19d66845ca02e378ff4/scipy/signal/filter_design.py#L2374

I'd suggest to add an explicit check for the type of fs and raise a specific exception in case of wrong types.

#### Reproducing code example:

```
scipy.signal.butter(1, 1, "highpass", fs=np.array([10, 20]))
```

#### Error message:

```
Traceback (most recent call last)
<ipython-input-5-0f771327b56e> in <module>
----> 1 scipy.signal.butter(1, 1, "highpass", fs=np.array([10, 20]))

~/python3.8/site-packages/scipy/signal/filter_design.py in butter(N, Wn, btype, analog, output, fs)
   2953     >>> plt.show()
   2954     """
-> 2955     return iirfilter(N, Wn, btype=btype, analog=analog,
   2956                      output=output, ftype='butter', fs=fs)
   2957 

~python3.8/site-packages/scipy/signal/filter_design.py in iirfilter(N, Wn, rp, rs, btype, analog, ftype, output, fs)
   2431     if btype in ('lowpass', 'highpass'):
   2432         if numpy.size(Wn) != 1:
-> 2433             raise ValueError('Must specify a single critical frequency Wn for lowpass or highpass filter')
   2434 
   2435         if btype == 'lowpass':

ValueError: Must specify a single critical frequency Wn for lowpass or highpass filter

```

#### Scipy/Numpy/Python version information:

1.6.2 1.20.2 sys.version_info(major=3, minor=8, micro=8, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by yetinam**
#### Reference issue

Closes #14450 

#### What does this implement/fix?

This PR adds input validation for the sampling frequency fs to the filters in signal.filter. Invalid inputs now raise an appropriate ValueError.

#### Additional information

My flake8 shows a list of code style issues in the files I touched. As far as I can tell I added no new code style issues, but I also didn't fix the existing ones.

### CODE REVIEW

File: scipy/signal/filter_design.py [hunk 1]
```diff
@@ -2027,6 +2043,7 @@ def bilinear(b, a, fs=1.0):
     >>> plt.ylabel('Magnitude [dB]')
     >>> plt.grid()
     """
+    _validate_fs(fs)
     fs = float(fs)
```
Comment (larsoner) [inline scipy/signal/filter_design.py:2047]: So at least some places it cannot be None, like here. I'd rather make this:
```suggestion
    fs = _validate_fs(fs)
```
and add a `allow_none=False` or `True` (depending on how common allowing `fs=None` is)

Comment (larsoner) [reply scipy/signal/filter_design.py:2047]: (and inside `_validate_fs`, if it's non-None, cast to float -- and in either case, return it)

Comment (yetinam) [reply scipy/signal/filter_design.py]: Hey @larsoner , thanks for the comments. I incorporated the explicit check for None and also added the cast to float and a return statement.

File: scipy/signal/filter_design.py [hunk 2]
```diff
@@ -56,6 +56,24 @@ def _is_int_type(x):
...
+    """
+    if not allow_none and fs is None:
+        raise ValueError("Sampling frequency can not be none.")
+    if fs is not None and not np.isscalar(fs):
+        raise ValueError("Sampling frequency fs must be a single scalar.")
+
+    if fs is None:
+        return None
+    else:
+        return float(fs)
```
Comment (larsoner) [inline scipy/signal/filter_design.py:74]: I would simplify the flow here a bit:
```suggestion
    if fs is None:
        if not allow_none:
            raise ValueError("Sampling frequency can not be none.")
    else:  # should be float
        if not np.isscalar(fs):
            raise ValueError("Sampling frequency fs must be a single scalar.")
        fs = float(fs)
    return fs
```

File: scipy/signal/fir_filter_design.py [hunk 3]
```diff
@@ -30,6 +30,24 @@ def _get_fs(fs, nyq):
...
+    """
+    if not allow_none and fs is None:
+        raise ValueError("Sampling frequency can not be none.")
+    if fs is not None and not np.isscalar(fs):
+        raise ValueError("Sampling frequency fs must be a single scalar.")
+
+    if fs is None:
+        return None
+    else:
+        return float(fs)
```
Comment (larsoner) [inline scipy/signal/fir_filter_design.py:48]: Having two copies of this violates DRY, I would put one copy in one place that both files can import from. `_arraytools` would be one option if you can't find a better one

File: scipy/signal/tests/test_filter_design.py [hunk 4]
```diff
@@ -827,6 +827,15 @@ def test_nyquist(self):
         assert_array_almost_equal(w1, w2)
         assert_array_almost_equal(h1, h2)
 
+    def test_fs_validation(self):
+        with assert_raises(ValueError) as e:
```
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:831]: better / more modern and readable form would be something like:
```suggestion
        with pytest.raises(ValueError, match='Sampling.*single scalar'):
```
and no need for the `assert_` later

File: scipy/signal/filter_design.py (hunk 2 above)
Comment (yetinam) [reply scipy/signal/filter_design.py]: Thanks for the suggestion. I incorporated it.

File: scipy/signal/fir_filter_design.py (hunk 3 above)
Comment (yetinam) [reply scipy/signal/fir_filter_design.py]: Thanks for pointing this out. I moved the function to _arraytools.

File: scipy/signal/tests/test_filter_design.py (hunk 4 above)
Comment (yetinam) [reply scipy/signal/tests/test_filter_design.py]: Thanks for pointing me to this syntax. It's integrated now.

**Comment by lucascolley** [comment]
It looks like all comments were addressed here. Sorry for the long wait @yetinam , would you like to merge main and resolve any conflicts? If so, we should be able to merge this some time soon.

**Comment by yetinam** [comment]
Hi @lucascolley ,
I had totally forgotten about this. I'll merge main and resolve the conflict in the next days.

**Comment by yetinam** [comment]
Turns out resolving all conflicts after 3.5 years of development is a bit of a hassle. I've created a new PR with identical changes instead in #19971.

**Pull request closed without merging**